core/arch/aarch64/emit_utils.c

```diff
@@ -574,19 +574,84 @@ void
 append_restore_simd_reg(dcontext_t *dcontext, instrlist_t *ilist, bool absolute)
 {
     int i;
-    /* add x1, x(dcxt), #(off) */
+    /* add x1, x(dcxt), #(offset simd) */
     APP(ilist,
         XINST_CREATE_add_2src(dcontext, opnd_create_reg(DR_REG_X1),
                               opnd_create_reg(REG_DCXT),
                               OPND_CREATE_INTPTR(offsetof(priv_mcontext_t, simd))));
     for (i = 0; i < 32; i += 2) {
-        /* ldp q(i), q(i + 1), [x1, #(i * 16)] */
+        /* ldp q(i), q(i + 1), [x1, #(i * 16)]
+         * From the AArch64 manual:
+         * "The signed immediate byte offset is a multiple of 16 in the range
+         * -1024 to 1008, defaulting to 0 and encoded in the imm7 field as
+         * <imm>/16."
+         */
         APP(ilist,
             INSTR_CREATE_ldp(
                 dcontext, opnd_create_reg(DR_REG_Q0 + i),
                 opnd_create_reg(DR_REG_Q0 + i + 1),
                 opnd_create_base_disp(DR_REG_X1, DR_REG_NULL, 0, i * 16, OPSZ_32)));
     }
+    if (proc_has_feature(FEATURE_SVE)) {
+        for (i = 0; i < 32; i++) {
+            /* ldr z(i), [x1, #(i mul vl)]
+             * From the SVE manual:
+             * "Load a vector register from a memory address generated by a
+             * 64-bit scalar base, plus an immediate offset in the range -256
+             * to 255 which is multiplied by the current vector register size
+             * in bytes."
```

Comment:
> But the IR level stores the absolute offset, so we need to multiply i by the vector size here. The IR level should not have to know how immediates/displacements might be encoded (as much as possible).

AIUI we do not need to multiply `i` by the vector size because SVE's unpredicated vector load (`LDR`) and store (`STR`) instructions interpret the immediate value as a vector index, not a byte offset or an element index.

From section 5.2.1.5 in the SVE reference manual at https://developer.arm.com/documentation/ddi0584/latest/
```
5.2.1.5 Unpredicated vector register load/store
The unpredicated vector register load, LDR, and store, STR, instructions transfer a single vector register from or to memory locations that are specified by a scalar base register plus an immediate index value that is in the range -256 to 255, inclusive. The immediate index value defaults to zero if omitted. For this group of SVE instructions:
- The immediate index value is a vector index, not an element index. The immediate index value is multiplied by the current vector register length in bytes.
. . .
Supported addressing modes            Assembler syntax
Scalar base + immediate index         [<Xn|SP>{, #<simm>, MUL VL}]
```
This is why the `MUL VL` keyword in SVE assembly is significant in such memory instructions. The instruction's micro-ops in the hardware will multiply with the supported vector length.

Reply:
But this breaks invariants about base-displacement operands: such an operand always refers to the linear address equal to the base register value plus the displacement.  There cannot be any opcode-specific adjustments secretly applied to the displacement: there would have to be a different type of operand if the displacement is modified by the opcode.  Otherwise this is going to break all the users of the IR such as drutil_insert_get_mem_addr() (and by extension all the address tracing tools built on DR) which assume they can examine a base-displacement operand in isolation from its containing instruction and compute its linear address.

I do not think it is worth creating some new kind of operand just for this case when there is a simple solution of the IR taking a byte offset as the displacement (i*vector-size here) and the decoder/encoder transform that into a vector index.

Reply:
To be clear, the proposed solution is for the IR to look and behave as the usual byte offset displacement with the vector-to-byte indexing transform done in the decode/encode functions. So for the SVE SIMD save/restore loops in `core/arch/aarch64/emit_utils.c`, this:
```
for (i = 0; i < 32; i++) {
. . .
    opnd_create_base_disp(DR_REG_X1, DR_REG_NULL, 0,
                          i,
                          opnd_size_from_bytes(proc_get_vector_length_bytes()))
. . .
```
changes to:
```
for (i = 0; i < 32; i++) {
. . .
    opnd_create_base_disp(DR_REG_X1, DR_REG_NULL, 0,
                          i * proc_get_vector_length_bytes(),
                          opnd_size_from_bytes(proc_get_vector_length_bytes()))
. . .
```
And the call to `opnd_create_base_disp()` has to change accordingly in the decode function `decode_opnd_svemem_gpr_simm9_vl()` as well as base+disp encoding in the encode function `encode_opnd_svemem_gpr_simm9_vl()`.

That would mean the disassembly tests in `suite/tests/api/dis-a64-sve.txt` would have to look like e.g.
```
85be5b82 : ldr z2, [x28, #10, mul vl] : ldr +0x140(%x28)[32byte] -> %z2
```
instead of the current:
```
85be5b82 : ldr z2, [x28, #10, mul vl] : ldr +0x0a(%x28)[32byte] -> %z2
```
for test vector length of 256 bits.

So this assembly language address:
```
[x28, #10, mul vl]
```
effectively means this in terms of IR:
```
[x28, #10 * vl]
```

So for a save/restore loop:
```
[x1]
[x1, #1 * vl]
[x1, #2 * vl]
[x1, #3 * vl]
. . .
[x1, #31 * vl]
```
If the above is a correct interpretation of the fix I can make the change in a separate PR as this one is already too large.

Reply:
Yes that all looks good.